trie/zk/merkle_tree.go

```
@@ -0,0 +1,482 @@
+	for i, node := 0, t.rootNode; i < t.maxLevels; i++ {
+		if i >= fromLevel {
+			// TODO: notice here we may have broken some implicit on the proofDb:
+			// the key is not kecca(value) and it even can not be derived from the value by any means without a actually decoding
+			if err := writeNode(node); err != nil {
+				return err
+			}
+		}
+		switch n := node.(type) {
+		case *ParentNode:
+			node = t.getChild(n, path.Get(i))
+			continue
+		case *LeafNode:
+		case *EmptyNode:
+		case *HashNode:
+		default:
+			return trie.ErrInvalidNodeFound
+		}
+		return nil
+	}
+	return nil
+}
+
+func (t *MerkleTree) checkNodeHashConflictDuringUpdate(node TreeNode) error {
+	if t.dirtyNode == nil {
+		return nil
+	}
+	return t.checkNodeHashConflict(node, true, nil)
+}
+
+func (t *MerkleTree) checkNodeHashConflict(
+	node TreeNode,
+	updating bool,
+	handleDirtyNode func(node TreeNode) error,
+) error {
+	if t.dirtyNode == nil {
+		return nil
+	}
+	return computeNodeHash(node, func(node TreeNode) error {
+		hashString := string(node.Hash()[:])
+		if oldNode, ok := t.dirtyNode[hashString]; ok {
+			if bytes.Equal(oldNode.CanonicalValue(), node.CanonicalValue()) {
+				return nil
+			}
+			return trie.ErrNodeKeyAlreadyExists
+		}
+		if updating {
+			// Compute the hash during tree modification.
+			// If you don't copy, even if the hash is changed many times and stored in the map,
+			// the pointer hasn't changed, so you'll always be left with the last state of the node.
+			t.dirtyNode[hashString] = copyNode(node)
+		} else {
+			t.dirtyNode[hashString] = node
+		}
+		if handleDirtyNode == nil {
+			return nil
+		}
+		return handleDirtyNode(node)
+	})
+}
+
+// getChild If the child node is a hash node, decode it and update the parent node.
+func (t *MerkleTree) getChild(node *ParentNode, path byte) TreeNode {
+	if t.findBlobByHash == nil {
+		return node.Child(path)
+	}
+	if hashNode, ok := node.Child(path).(*HashNode); ok {
+		blob, _ := t.findBlobByHash(hashNode.Hash()[:])
```

Why do you ignore the err here?

---

because not finding the blob is not always an error case: if we don't find the blob, it will remain a hash node and the function using getChild will take care of it.
If I were to take a log there, it would likely be a lot of logs.

---

Could you please a comment of this?